### ISSUE TITLE
Area to pixels conversion functions

### ISSUE THREAD

**Issue by pnorman** [maplibre/maplibre-gl-js#1925]
## Motivation

It is common to want to style or filter point labels based on the number of pixels their corresponding area takes up. The area in meters can be exposed through the vector tiles, and is commonly calculated in web mercator square meters by data loaders like osm2pgsql and imposm.

In Mapnik raster tiles, it is possible to convert this value to pixels, but this is not possible with vector tiles where overzoom is used.

## Design Alternatives

Doing nothing makes it difficult to do this styling without doing math that is easy to get wrong.

An alternative design would be to have a pixels to meters function, which would work equally well.

## Design

A math function in the style language that converts a number (in square meters in web mercator) into a number of pixels. An accompanying function to convert a number (in mercator meters) to a length in pixels would also be good.

### Mock-Up

- `["area2pixels", number]` to convert `number` into pixels at the current `[zoom]`
- `["length2pixels", number]` to convert `number` into pixels at the current `[zoom]`

Example usage would be
```json
{"filter": [">=", ["area2pixels", ["get", "way_area"]], 750]}
```

The current way to do this would be

```json
{"filter": [">=", ["get", "way_area"], ["*", 750, 6126430366.1, ["^", 0.25, ["zoom"]]]]
```
#### How will we teach this design?
It will be documented in the expressions documentation

#### What existing precedents support the new concepts?
Use of way_pixels by Mapnik raster styles

### Implementation

To be implemented similarly to other math functions.

**Comment by HarelM** [comment]
I think this can be a good addition.
There are some questions to think about regarding implementation:
What is the input? the feature id? the feature geometry?
Can this cause performance degradation?
Do we calculate it for every feature and cache it?
Shouldn't we pre-process this data into the tiles and simply have a simple conversion between length/area in meters to pixels?

Overall, I think this can be a good improvement if done correctly.

**Comment by pnorman** [comment]
> What is the input? the feature id? the feature geometry?

No. As established above, the input is a number which is an area in square meters.

> Can this cause performance degradation?

I  can't see how it would - are there performance issues with the other math features?

> Do we calculate it for every feature and cache it?

This would make no sense, as it's not a function that operates on a feature, but on a number.

> Shouldn't we pre-process this data into the tiles and simply have a simple conversion between length/area in meters to pixels?

This is a simple conversion between meters to pixels.

**Comment by HarelM** [comment]
Thanks for the info, I think this is more clear to me now.
If this can be achieved right now, (with a long expression, I know, as you wrote above) what's the motivation to create a shorter method "limited" method?

**Comment by pnorman** [comment]
It's a shorter method that makes styles clearer and involves no magic numbers in the style, similarly to how the multiple logarithm math functions make it simpler.

**Comment by HarelM** [comment]
I'll bring it up I the next steering committee meeting and see what people think.

**Comment by ZeLonewolf** [comment]
This would be very helpful for coming up with size-appropriate labels for area features.

**Comment by HarelM** [comment]
I'm sorry to say, but the TSC had decided that this feature is not wanted at this point. It can be achieved using the current syntax and will require a rework when supporting more projections.
Sorry.

**Comment by johncarmack1984** [comment]
The TSC declined this on 2022-12-17. This one can be closed.

**Comment by HarelM** [comment]
Feel free to join the following design discussion:
- https://github.com/maplibre/maplibre-style-spec/issues/1881
